drivers/hmis/app/models/hmis/hud/processors/base.rb

```diff
@@ -73,4 +80,53 @@ def factory_name
   def schema
     raise 'Implement in sub-class'
   end
+
+  def process_custom_field(field, value)
+    record = @processor.send(factory_name)
+    return false unless record.respond_to?(:custom_data_elements)
+
+    cded = Hmis::Hud::CustomDataElementDefinition.for_type(record.class.name).where(key: field).first
+    return false unless cded.present?
+
+    attrs = {
+      user: @processor.hud_user,
+      data_source_id: @processor.hud_user.data_source_id,
+      data_element_definition: cded,
+    }
+    value_field_name = "value_#{cded.field_type}"
+    value = attribute_value_for_enum(nil, value) # converts HIDDEN => nil
+
+    existing_values = record.custom_data_elements.where(data_element_definition: cded, owner: record)
+
+    # If this custom field only allows 1 value and there already is one, update it.
+    if !cded.repeats && existing_values.exists?
+      cde_attributes = { id: existing_values.first.id }
+      if value.present?
+        cde_attributes[value_field_name] = value
+        cde_attributes.merge!(attrs)
+      else
+        cde_attributes[:_destroy] = 1
+      end
+
+      record.assign_attributes(
+        custom_data_elements_attributes: [
+          # Update or delete the existing value
+          cde_attributes,
+          # Delete any other values (there shouldn't be any, but just in case)
+          *existing_values.drop(1).map { |old_cde| { id: old_cde.id, _destroy: 1 } },
+        ],
+      )
+    # Else create new custom field value(s), and delete any existing ones.
+    else
+      record.assign_attributes(
+        custom_data_elements_attributes: [
+          # Add new value(s)
+          *Array.wrap(value).map { |new_value| { value_field_name => new_value, **attrs } },
+          # Destroy any existing values for this custom field
+          *existing_values.map { |old_cde| { id: old_cde.id, _destroy: 1 } },
```

Comment:
Is there any point in trying to preserve the old values if they haven't changed?  Seems like this approach could lead to a bunch of churn in the custom table....

Reply:
The churn would really only happen for custom fields that allow multiple values. If it only allows one value, we just update it if it exists (which might be a no-op). But still, I think it's worth avoiding that churn.

Reply:
https://github.com/greenriver/hmis-warehouse/pull/3065/commits/c1542875989bc901041df37fecad88c7c332ff4a